added new MultiStream operator: ExtractFieldValues

dafnapension opened the pull request:
… FilterByCondsOnValues
per issue 297, and a demand from the field..

dafnapension commented:
Hi @yoavkatz , I added, at last, the Class that we talked about and you summarized in issue #297.
this is FindAndStoreMostCommonValues.
'on the way' I added yet another Class that extends FilterByValue to a more general condition for filtering: FilterByCondsOnValues.  A need for this came from some application where the language of the document is hiding inside a field, rather than being the whole field, and yet, we need to filter by that language.
The tests show expected results for some simple examples.
please review.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+        stream = multi_stream[self.stream_name]
+        all_values = []
+        for instance in stream:
+            all_values.append(instance[self.field_name])
```
Need to support list elements

dafnapension replied:
As I wrote above: not sure if to support them individually: count the frequency of each individual, and then which records are to stay? those whose lists consist of all frequent values? whose lists contain at least one frequent value? 
Or rather: measure frequencies of whole lists, and then the same rule used for individual values would apply for whole lists, and the issue is just to allow, programmatically, lists of str, and not only   str.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+    Do nothing (exception thrown) if the multistream does not contain a stream named stream_name, or if the multistream does not
+    contain a field (feature) named field_name.
+    Count the frequency of each value, over the stream named stream_name. Sort the values by frequency.
+    Discard each value that is not at the top_percent of that sorted list.
```
Not sure if there is a reason to take first 80% of the list. One can always cut it later. Maybe I'm missing something.

dafnapension replied:
I accept. I now understand what you had in mind: measure only by frequency of coverage.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+        min_frequency = math.floor(self.min_relative_freq_percent * len(all_values) / 100)
+        while top_values[-1][1] < min_frequency:
+            top_values.pop()
+        values_to_remain = [ele[0] for ele in top_values]
```
values_to_keep?

dafnapension replied:
yes, better. Thanks!

yoavkatz reviewed (changes requested):
Thanks Dafna - added comments in the PR.  If you want to can also talk. Not sure all my comments were clear.

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
                 yield instance
 
 
+class FilterByCondsOnValues(SingleStreamOperator):
```
And where will the value computed in the first stage be kept for the second stage? extend the record with yet another field?

yoavkatz replied:
Apply can receive any function, but I think it needs to be named function. It stores the results in a new field or the same field.

I'll update the doc, but here is a copy:

  """A class used to apply a python function and store the result in a field.

    Args:
        function (str): name of function. 
        to_field (str): the field to store the result
        additional arguments are field names passed to the function

    Examples:

    Store in field  "b" the uppercase string of the value in field "a"
    Apply("a", function=str.upper, to_field="b")   

    Dump the json representation of field "t" and store back in the same field.
    Apply("t", function=json.dumps, to_field="t")

    Set the time in a field 'b'.
    Apply(function=time.time, to_field="b")

dafnapension replied:
Thanks, @yoavkatz ,
If the argument to the passed function also needs to be passed upon creation of Apply, then you need to create one Apply object per record, and not one Apply per stream?

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+
+class FindAndStoreMostCommonValues(MultiStreamOperator):
+    field_name: str
+    stream_name: Optional[str] = "train"
```
I accept.

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+class FindAndStoreMostCommonValues(MultiStreamOperator):
+    field_name: str
+    stream_name: Optional[str] = "train"
+    top_percent: Optional[int] = 80
```
I accept

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+    min_relative_freq_percent: Optional[int] = 5
+    new_field: Optional[str] = None
+    """
+    Given a MultiStream, a stream_name, and a field name, and two margins of commonality.
```
Very nice. I accept. I now understand what you had in mind: both measures deal with the relative frequency of the unique values (this, I now understand, is what you had in mind) and not their relative rank in the sorted list of relative frequencies (as I first implemented).

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+                yield instance
+
+
+class FindAndStoreMostCommonValues(MultiStreamOperator):
```
Yes, better. I accept.

dafnapension commented on src/unitxt/operators.py:
```diff
@@ -671,6 +672,69 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+    new_field: Optional[str] = None
+    """
+    Given a MultiStream, a stream_name, and a field name, and two margins of commonality.
+    Do nothing (exception thrown) if the multistream does not contain a stream named stream_name, or if the multistream does not
```
I replaced my description with your much nicer one.  Thanks!

dafnapension commented:
Thank you very much, @yoavkatz , for your thorough review. I got it all, and replied above accordingly, 
except two things:  
(1) I did not understand the semantics of  `process_every_value=True` in `ExtractFieldValues`,
(2) And was not sure where to write the computed value, from the first stage that you suggested, of `FilterByCondsOnValues`.

dafnapension commented:
Thank you very much, @yoavkatz , for your thorough review. I implemented, and ran tests, all your comments above (which I also replied to), except two things:
(1) I did not understand the semantics of `process_every_value=True` in `ExtractFieldValues`,
(2) And I was not sure where to write the computed value, from the first stage that you suggested, of `FilterByCondsOnValues`.

dafnapension commented:
Hi @yoavkatz , 
I think I got it my question (1) above:  here is how I am going to implement:
When the value of 'field' is list, and not a single value:
if `process_every_value=False`,  I treat the list as one element, and like before, find the most popular lists, counting the number of occurrences of each and every unique list in column 'field'. At the end, I also report in the 'to_field'  -- a list of lists, being the most popular lists.
if `process_every_value=True`, I count the occurrences of each and every individual element in that list.  e.g., the list ['a', 'b', 'c'], showing in field 'field' of a row in 'train' (say), contributes 1 to the number of occurrences of 'a',  1 to the number of occurrences of 'b' and 1 to the number of occurrences of 'c'.  At the end I report the most popular individual **values**.

Later, down the stream of implementation, people will do with the values they find in field 'to_field' -- as they want. Either filter out any row hose list contains no popular element, or filter out any row whose list is not wholly made of popular elements.

dafnapension commented:
Hi @yoavkatz ,
I implemented as I understood and described above.
If this is in the direction of what you had in mind - then I am only left with the question: where to write the value that you suggested [here](https://github.com/IBM/unitxt/pull/316#discussion_r1398409893).

yoavkatz commented on tests/test_operators.py:
```diff
@@ -375,6 +399,336 @@ def test_merge(self):
+                "test": [{"field": "test1"}],
+                "validation": [{"field": "validation1"}],
+                "train": [
+                    {"field": "train1"},
```
I suggeste make the test more interesting (and test more of the behavior). Add 10 examples with more meaningful names, not in order of frequency from the start.

fish
dog
dog
cat
dog
cat
sheep
cat
fish
shark





tain

dafnapension replied:
Good idea,  will do.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -671,6 +674,100 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+            while values_and_counts[-1][1] < min_frequency:
+                values_and_counts.pop()
+        values_to_keep = [[*ele[0]] if isinstance(ele[0], tuple) else ele[0] for ele in values_and_counts]
+        if self.to_field is None:
```
I suggest we just throw an error if to_field is not set.  It's more consistent with other operators.

dafnapension replied:
I accept.

dafnapension commented:
Thank you, @yoavkatz , for your very good suggestions. I will implement them.

dafnapension commented:
Hi @yoavkatz, I implemented changes that you asked.
Thank you for your thorough and educating review.

dafnapension commented:
Hi @yoavkatz ,
My joy about Apply was too early.. I think the method is bugged.
As you can see in PR https://github.com/IBM/unitxt/pull/321, I tried to use it with a simple function I defined in the module almostEvilML_qa_by_lang.  That function is translated to a string by Apply, and stored in the cards that show in src/catalog/cards/almostEvilML_qa_by_lang.
However, the translation to string is miserable and erroneous, and hence on loading time, the function can not be restored, and hence the tests fail when I push.

This is irrelevant to this PR, but since we talked about Apply in this PR, I am writing this comment..

yoavkatz commented:
> Hi @yoavkatz , My joy about Apply was too early.. I think the method is bugged. As you can see in PR #321, I tried to use it with a simple function I defined in the module `almostEvilML_qa_by_lang`. That function is translated to a string , `"__main__.extract_lang_from_METADATA"`, by Apply, and stored in the cards that show in src/catalog/cards/almostEvilML_qa_by_lang.
> 
> It seems, though, that global lists, `__builtins__` and `globals()` are not updated as necessary, and the restoring of the function from its string format - fails.
> 
> This is irrelevant to this PR, but since we talked about Apply in this PR, I am writing this comment here

Right.   We can not use functions which are not in the main code base of unitxt or fm-eval, because it will have no way to find them in the prepare cards.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -677,6 +680,74 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+                if sum_counts >= top_frequency:
+                    break
+            values_and_counts = counter.most_common(i + 1)
+        if self.min_frequency_percent > 0:
```
Thinking about this more.  I think we should not support both top_frequency_percent and min frequency_precent together, since if you take min_frequency=20, you might not be able to get top_frequency_percent.  

I suggest that for now, we check this and throw an error if both are set.  
  This can be done in the verify method, which is called before prepare.

yoavkatz replied:
We should also verify both numbers are between 0 and 100.

dafnapension replied:
Done. Very nice catch!

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -677,6 +680,74 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+        stream = multi_stream[self.stream_name]
+        all_values = []
+        for instance in stream:
+            if (not isinstance(instance[self.field], list)) or (self.process_every_value == False):
```
IS this the logic?
  if (self.process_every_value):
     if (not isinstance(instance[self.field], list)) 
         raise ValueError(...)
     all_values.extend(instance[self.field])
  else
      all_values.append(instance[self.field])
      
      What Am I missing?  Why do this " (*instance[self.field],) if isinstance(instance[self.field]," for lists?

yoavkatz commented on tests/test_operators.py:
```diff
@@ -375,6 +377,282 @@ def test_merge(self):
+            {
+                "test": [{"field": ["a", "b", "c"]}],
+                "validation": [{"field": ["d", "e", "f"]}],
+                "train": [
```
```suggestion

 # The list has 
 # h: 6
  # m: 6
 # j: 3
 # i: 3 
 # k: 3
 # o: 3 
 # p: 3 
  # q: 2
 # r: 2
 # s: 2
 # t: 1
 # u:1 
 # v:1
 
# Tuples:

# ["h", "I", "j"] : 3
# ["k", "h", "m"]: 3
# ["m", "n", "o"] : 3
 # ["q", "r", "s"] : 2
 # ["t","u","v"] : 1

  
                "train": [
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -677,6 +680,74 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+        stream = multi_stream[self.stream_name]
+        all_values = []
+        for instance in stream:
+            if (not isinstance(instance[self.field], list)) or (self.process_every_value == False):
```
I think that if we process_every_value = True and the value is not a list, an error should be generated.

dafnapension replied:
Thanks! indeed important!  Done.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -677,6 +680,74 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
+        all_values = []
+        for instance in stream:
+            if (not isinstance(instance[self.field], list)) or (self.process_every_value == False):
+                all_values.append(
```
Add a comment why we convert the list to a tuple (took me a while to understand).

I think this is the logic:
1. If process_every_value and value is not a list, throw an error.
2. If not process_every_value and is not a list, append element to the all_values list.
3. If not process_every_value and list,  convert to tuple, and append tuple to all_values list.
4. If process_every_value and list, add each element in list to the all_values list.

dafnapension replied:
Exactly! conversion is done only to allow the simple use of Counter, which would not accept lists hashables - for its entries.

yoavkatz reviewed (changes requested):
Thank Dafna.  I had another review - hopefully the final one , with a few comments, mostly on documentation and error checking.

dafnapension commented:
Thank you, @yoavkatz , I addressed your comments. Please do not hesitate to further review and write comments. They help me improve my contribution, and I learn a lot from them.

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -695,39 +695,62 @@ class ExtractFieldValues(MultiStreamOperator):
     ExtractFieldValues(stream_name="train", field="label", to_field="classes",overall_top_frequency_percent=80) -
-    extracts the most frequent possible values of field 'label' that cover at least 80% of the samples,
-    and stores them in field 'classes' of all streams.
+    extracts the most frequent possible values of field 'label' that cover at least 80% of the rows of stream_name,
```
```suggestion
    extracts the most frequent possible values of field 'label' that cover at least 80% of the instances of stream_name,
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -695,39 +695,62 @@ class ExtractFieldValues(MultiStreamOperator):
-    extracts the most frequent possible values of field 'label' that cover at least 80% of the samples,
-    and stores them in field 'classes' of all streams.
+    extracts the most frequent possible values of field 'label' that cover at least 80% of the rows of stream_name,
+    and stores them in field 'classes' of each record of all streams.
```
```suggestion
    and stores them in field 'classes' of each instance in all the streams.
```

yoavkatz commented on src/unitxt/operators.py:
```diff
@@ -695,39 +695,62 @@ class ExtractFieldValues(MultiStreamOperator):
+        ), "'min_frequency_percent' must be between 0 and 100"
+        assert not (
+            self.overall_top_frequency_percent < 100 and self.min_frequency_percent > 0
+        ), "At most one of 'overall_top_frequency_percent' and 'min_frequency_percent' is allowed to move from their default value"
```
```suggestion
        ), "At most one of 'overall_top_frequency_percent' and 'min_frequency_percent' is allowed to be set "
```

yoavkatz reviewed (approved):
Minor two suggestions  regarding teminology. We have "streams" that contain "instances" (not rows or records)

dafnapension commented:
Hi @yoavkatz ,
I thought about your question regarding the complex condition around what to append/extend to all_values in `process(..)`, and thought that it is not too efficient to check these for each and every instance, as all the instances have the same 'answers' to these condition.
So I moved these condition-evaluation to out of the loop.
Could you please review this as well?  (pardonez moi)
(Like in the old joke of my grandmother: when at last the old maid is invited to a dance, the little orchestra goes out for a pee recess..)

yoavkatz commented:
> Hi @yoavkatz , I thought about your question regarding the complex condition around what to append/extend to all_values in `process(..)`, and thought that it is not too efficient to check these for each and every instance, as all the instances have the same 'answers' to these condition. So I moved these condition-evaluation to out of the loop. Could you please review this as well? (pardonez moi) (Like in the old joke of my grandmother: when at last the old maid is invited to a dance, the little orchestra goes out for a pee recess..)

I think what was before is safer, because it could be that fields are of different values (likely to be a bug, but still better to identify it).  Also in terms of runtime - we iterato over the whole stream, so adding another check is minimal cost.

dafnapension commented:
Thank you @yoavkatz , nice catch. Did not occur to me.   I am the little orchestra (KleiZmer) that went out to a pee recess.  Returning to the dancing hall right away.
